src/sbvr-api/control-flow.ts

```diff
@@ -0,0 +1,61 @@
+import * as _ from 'lodash'
+import * as Promise from 'bluebird'
+
+// lifts f to map over the changeSet if present
+export const liftP = <T, U>(fn: (v: T) => U | Promise<U>) => {
+	return ((a: T | T[]) => {
+		if (_.isArray(a)) {
+			// This must not be a settle as if any operation fails in a cs
+			// we want to discard the whole
+			return Promise.mapSeries(a, fn)
+		} else {
+			return Promise.resolve(a).then(fn)
+		}
+	// TODO: We cast here to merge `Promise<U> | Promise<U[]>` into `Promise<U | U[]>,
+	// which is equivalent for our uses but fixes issues where typing was being rejected
+	}) as ((a: T | T[]) => Promise<U | U[]>)
+}
+
+interface MappingFunction {
+	<T, U>(a: T[], fn: (v: T) => U | U[] | Promise<U>): Promise<Array<U | U[] | Error>>
+}
+
+// The settle- versions of
+const settleMapSeries: MappingFunction = (a, fn) => {
+	const runF = Promise.method(fn)
+	return Promise.mapSeries(a, _.flow(runF, wrap))
+}
+
+// Wrap a promise with reflection. This promise will always succeed, either
+// with the value or the error of the promise it is wrapping
+const wrap = <T>(p: Promise<T>) => {
+	return p.then(_.identity as (value: T) => T, _.identity as (value: Error) => Error)
+}
+
+// Maps fn over collection and returns an array of Promises. If any promise in the
+// collection is rejected it returns an array with the error, along with all the
+// promises that were fulfilled up to that point
+const mapTill: MappingFunction = <T, U>(a: T[], fn: (v: T) => U) => {
+	const runF = Promise.method(fn)
+	const results: Array<U | any> = []
+	return Promise.each(a, (p) => {
+		return runF(p)
+		.then((result) => {
```

Comment:
This is actually giving me an implicit `any` type error. I think the problem is with the signature in `Promise.method`. I wrapped `Promise.method` in 
```
function toPromiseFn<T> (fn:(...args: any[]) => (T | Promise<T>)) {
	return Promise.method(fn) as (...args: any[]) => Promise<T>
}
```
This way, it will correctly infer the type of result in here

Reply:
For me `result` is typed as `U`, can you check you've got all the types installed?

Reply:
looks like the problem is with the required bluebird version. If I update to 3.5.16 it infers like yours, but the problem is that I had 3.5.4 installed previously which satisfied the constraint in package.json but made it fail.

Reply:
I've updated package.json to the latest versions to help a bit with with this